x-pack/plugins/observability_solution/observability_ai_assistant/public/hooks/use_conversation_key.ts

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useRef } from 'react';
+import { v4 } from 'uuid';
+
+// Regenerate the key only when the id changes, except after
+// creating the conversation. Ideally this happens by adding
+// state to the current route, but I'm not keen on adding
+// the concept of state to the router, due to a mismatch
+// between router.link() and router.push(). So, this is a
+// pretty gross workaround for persisting a key under some
+// conditions.
+
+export function useConversationKey(conversationId: string | undefined) {
+  const keyRef = useRef(v4());
+  const conversationIdRef = useRef(conversationId);
+
+  if (conversationIdRef.current !== conversationId) {
+    keyRef.current = v4();
+  }
+
+  conversationIdRef.current = conversationId;
+
+  return {
+    updateConversationIdInPlace: (nextConversationId: string) => {
+      conversationIdRef.current = nextConversationId;
+    },
+    key: keyRef.current,
+  };
+}
```

Comment:
Not sure if this is gross. But it definitely is confusing. Afaiu this is to make sure that the conversation key is updated when the conversation id updates, in order to refresh it in the list. But I'm still not 100% sure why it has to be this convoluted (a hook with two refs that needs to be manually updated via `updateConversationIdInPlace`).

Reply:
there is a comment above, does it need extra clarification?

Reply:
it's not about the list, but about updating the url from /new to /:id without unmounting

Reply:
> it's not about the list, but about updating the url from /new to /:id without unmounting

Ahhh! I missed that. So we actually perform a redirect but keep the DOM tree untouched. That's nice but could be clearer (maybe I am just blind)